life360/life360_app.groovy

```diff
@@ -321,303 +377,470 @@ def listCircles() {
+            ]
+            asynchttpPost("ackHandler", requestParams, [:])
+
+            /*
+			addHttpR(postUrl)
+			def requestResult; requestResult = null
+			Integer rc
+			httpPost(uri: postUrl, body: ["type": "location"], headers: ["Authorization": "Bearer ${state.life360AccessToken}"]) {response ->
+				rc = response.getStatus()
+				requestResult = response
+			}
+			def requestId = requestResult.data?.requestId
+			Boolean isPollable = requestResult.data?.isPollable
+			if (getSettingB('logEnable')) {
+				debug "PlaceHandler Post rc: $rc response.data = ${requestResult?.data}"
+				debug "PlaceHandler Post requestId = ${requestId} isPollable = $isPollable"
+			}
+			remTsVal('lastMembersDataUpdDt')
+			runIn(2, 'schedUpdateMembers')
+			*/
+
+        } catch (e) {
+            logError "request post / get, error: ", e
+        }
     }
-  
-    // we got a PUSH EVENT from Life360 - better update everything by pulling a fresh data packet
-    // But first, wait a second to let packet catch-up with push event
-    updateMembers()
 }
 
+void ackHandler(resp, data) {
+	Integer rc = resp.getStatus()
+	if(rc == 200 || rc == 207) {
+		def result = resp.getJson()
+		def requestId = result?.requestId
+		Boolean isPollable = result?.isPollable
+		if (getSettingB('logEnable')) {
+			debug "PlaceHandler Post response = ${result}"
+			debug "PlaceHandler Post requestId = ${requestId} isPollable = $isPollable"
+		}
+	} else {
+        // on 401 Unauthorized response, clear access token
+        if (rc == 401) {
+            state.life360AccessToken = null
+        }
+	   logWarn "response failed rc: $rc"
+	}
+
+	// we got a PUSH EVENT from Life360 - better update everything by pulling a fresh data packet
+	// But first, wait a second to let packet catch-up with push event
+	remTsVal('lastMembersDataUpdDt')
+	runIn(2, 'schedUpdateMembers')
+}
 
 def refresh() {
-    if (logEnable) log.debug("Life360+: refresh:")
-    listCircles()
-    updateMembers()
-    scheduleUpdates()
+	if (getSettingB('logEnable')) debug("refresh:")
+
+	Integer lastUpd = getLastTsValSecs('lastRefreshUpdDt')
+	if (lastUpd < 20) {
+		if (getSettingB('logEnable')) logInfo "Skipping circle sub update, last done $lastUpd"
+		return
+	}
+	updTsVal('lastRefreshUpdDt')
+
+	listCircles()
+	createCircleSubscription()
 }
 
-def scheduleUpdates() {
-    unschedule()
+void scheduleUpdates(Boolean reset = false) {
+    Boolean le = getSettingB('logEnable')
+
+	Integer lastRefreshSecs = state.lastRefreshSecs
+	Integer refreshSecs; refreshSecs = null
+	if (getSettingStr('pollFreq') == "auto") {
+		// adjust refresh rate based on if devices are moving
+		Integer numLocationUpdates = state.numLocationUpdates != null ? (Integer)state.numLocationUpdates : 0
+		// TODO: experiment with these values.. make sure we're not calling the API too frequently but still get timely user updates
+		if (numLocationUpdates == 0) {
+			// update every 300 (was 30,15,10) seconds
+			refreshSecs = 300
```

Comment:
whoa.. this isn't good. We should never be updating members more frequently than 1 minute and when active every 10 seconds is what I wrote.

I didn't see this originally but now I need to take control of this again until I have more time